client/network/sync/src/block_announce_validator.rs

```diff
@@ -0,0 +1,484 @@
+use sp_consensus::block_validation::Validation;
+use sp_runtime::traits::{Block as BlockT, Header, Zero};
+use std::{
+	collections::{hash_map::Entry, HashMap},
+	pin::Pin,
+	task::{Context, Poll},
+};
+
+/// Log target for this file.
+const LOG_TARGET: &str = "sync";
+
+/// Maximum number of concurrent block announce validations.
+///
+/// If the queue reaches the maximum, we drop any new block
+/// announcements.
+const MAX_CONCURRENT_BLOCK_ANNOUNCE_VALIDATIONS: usize = 256;
+
+/// Maximum number of concurrent block announce validations per peer.
+///
+/// See [`MAX_CONCURRENT_BLOCK_ANNOUNCE_VALIDATIONS`] for more information.
+const MAX_CONCURRENT_BLOCK_ANNOUNCE_VALIDATIONS_PER_PEER: usize = 4;
+
+/// Item that yields [`Stream`] implementation of [`BlockAnnounceValidator`].
+#[derive(Debug, Clone, PartialEq, Eq)]
+pub(crate) enum BlockAnnounceValidationResult<H> {
+	/// The announcement failed at validation.
+	///
+	/// The peer reputation should be decreased.
+	Failure {
+		/// The id of the peer that send us the announcement.
+		peer_id: PeerId,
+		/// Should the peer be disconnected?
+		disconnect: bool,
+	},
+	/// The announcement was validated successfully and should be passed to [`crate::ChainSync`].
+	Process {
+		/// The id of the peer that send us the announcement.
+		peer_id: PeerId,
+		/// Was this their new best block?
+		is_new_best: bool,
+		/// The announcement.
+		announce: BlockAnnounce<H>,
+	},
+	/// The block announcement should be skipped.
+	Skip {
+		/// The id of the peer that send us the announcement.
+		peer_id: PeerId,
+	},
+}
+
+impl<H> BlockAnnounceValidationResult<H> {
+	fn peer_id(&self) -> &PeerId {
+		match self {
+			BlockAnnounceValidationResult::Failure { peer_id, .. } |
+			BlockAnnounceValidationResult::Process { peer_id, .. } |
+			BlockAnnounceValidationResult::Skip { peer_id } => peer_id,
+		}
+	}
+}
+
+/// Result of [`BlockAnnounceValidator::allocate_slot_for_block_announce_validation`].
+enum AllocateSlotForBlockAnnounceValidation {
+	/// Success, there is a slot for the block announce validation.
+	Allocated,
+	/// We reached the total maximum number of validation slots.
+	TotalMaximumSlotsReached,
+	/// We reached the maximum number of validation slots for the given peer.
+	MaximumPeerSlotsReached,
+}
+
+pub(crate) struct BlockAnnounceValidator<B: BlockT> {
+	/// A type to check incoming block announcements.
+	validator: Box<dyn sp_consensus::block_validation::BlockAnnounceValidator<B> + Send>,
+	/// All block announcements that are currently being validated.
+	validations: FuturesUnordered<
+		Pin<Box<dyn Future<Output = BlockAnnounceValidationResult<B::Header>> + Send>>,
+	>,
+	/// Number of concurrent block announce validations per peer.
+	validations_per_peer: HashMap<PeerId, usize>,
+	/// Wake-up event when new validations are pushed.
+	event: Event,
+	/// Listener for wake-up events in [`Stream::poll_next`] implementation.
+	event_listener: Option<EventListener>,
```

Comment:
Could you explain why this is needed? I read your comment and the stream is not getting woken up but why is it necessary to add this machinery to make it wake up?

By the looks of it, a block announcement is pushed if we receive a notification and at the end of `SyncingEngine::poll()`, the block announcement stream is polled so by my reasoning the newly added announcement should be polled as well (albeit not explicitly but along with any other block announcement) but this is not the case, why?

Reply:
You are correct, in our implementation of `SyncingEngine::poll()` futures unordered are always polled after pushing a block announcement. But I decided to make `BlockAnnounceValidator` stream future-proof by ensuring the task is woken up even if the block announcements are pushed from somewhere else, and not before polling the stream. We can eliminate this and simplify the implementation, but then we would need to write warnings in the comments stating that the `BlockAnnounceValidator` stream must be explicitly polled after pushing a block announcement.

Reply:
Would any of this be an issue in the first place of `SyncingEngine` would use `tokio::select!` for polling all of its futures simultaneously? That's what we ultimately want and AFAIU and it would also give the possibility of pushing a block announcement anywhere in the and still getting polled.

Reply:
Yes, in case we use proper `.await` and `select!` this can introduce an unneeded latency in block announce validation: the newly added block announcement might not be polled until an already added one resolves. I will look what can be done with this.

Reply:
>the newly added block announcement might not be polled until an already added one resolves

I don't understand this. If the block announcements are stored in `FuturesUnordered` and they're getting polled as a group, I don't understand why some block announcement wouldn't get polled unless another one resolved. Am I overlooking some detail in the implementation?

Reply:
Yes, this is the property of `FuturesUnordered`:
> When new futures are added, [poll_next](https://docs.rs/futures/latest/futures/stream/trait.Stream.html#tymethod.poll_next) must be called in order to begin receiving wake-ups for new futures.

So, if there are some futures already and the task is waiting, `FuturesUnordered::poll_next` might not be called until some of them wakes the task up. Only after that the newly added feature will be registered in the task.

Reply:
@altonen Could you take a look again, please? Now the code must be ready for pushing block announcements from a different place and just `select!`-ing on `block_announce_validator.next()` in `SyncingEngine`.

Reply:
I'm not super happy with this approach either because I feel as though it complicates a relatively simple problem (push a future to `FuturesUnordered` and poll that container to get next ready future) but I understand that with the current architecture in `SyncingEngine` it may not be possible. The code itself looks fine but I still think we can improve this polling in the future so could you create an issue for it?

This PR probably has to be migrated to polkadot-sdk since Anton's review has gone stale